### ISSUE TITLE
Linking errors with Emscripten latest-upstream (LLVM 9)

### ISSUE THREAD

**Issue by akien-mga** [godotengine/godot#30270]
**Godot version:**
3.2 master (266dc732)

**OS/device including version:**
Mageia 7 x86_64
Emscripten 1.38.30, using LLVM 9 upstream backend (`latest-upstream`)

**Issue description:**
Trying the `latest-upstream` branch of Emscripten which is about to become the new default backend. Doing a `release` build.

There's a build issue (#30248), but using [that patch](https://github.com/godotengine/godot/issues/30248#issuecomment-507813049) it can be bypassed to reach the issue described here.

Then two linking issues are raised:

- [x] Duplicate symbol: `DefaultControllerMappings::mappings`

```
Linking Static Library ==> main/libmain.javascript.opt.bc
wasm-ld: error: duplicate symbol: DefaultControllerMappings::mappings
>>> defined in main/default_controller_mappings.gen.javascript.opt.bc
>>> defined in main/default_controller_mappings.gen.javascript.opt.bc
shared:ERROR: '/home/akien/Projects/godot/emscripten/emsdk/upstream/bin/wasm-ld -o main/libmain.javascript.opt.bc --allow-undefined --import-memory --import-table --lto-O0 main/default_controller_mappings.gen.javascript.opt.bc main/input_default.javascript.opt.bc main/main.javascript.opt.bc main/main_timer_sync.javascript.opt.bc main/performance.javascript.opt.bc main/default_controller_mappings.gen.javascript.opt.bc -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --export __wasm_call_ctors --export __data_end --export main --export malloc --export free --export setThrew --export __errno_location --export fflush -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024 --relocatable' failed (1)
```

I've looked at the `DefaultControllerMappings` code and I don't see why `wasm-ld` thinks it has a duplicate symbol. The symbol is declared in the header as `static const char *mappings[];` and implemented once in the .cpp. The header guards seem correct too.

- [x] `BINARYEN_TRAP_MODE` is not supported by the LLVM wasm backend

```
Linking Program        ==> bin/godot.javascript.opt.js
shared:ERROR: BINARYEN_TRAP_MODE is not supported by the LLVM wasm backend
```

It can simply be disabled with the following patch. I haven't checked yet what it does, if it's worth keeping on non-LLVM and if so how to handle that in `detect.py`:
```diff
diff --git a/platform/javascript/detect.py b/platform/javascript/detect.py
index c6afa02c6..f0191a71b 100644
--- a/platform/javascript/detect.py
+++ b/platform/javascript/detect.py
@@ -123,7 +123,7 @@ def configure(env):
     ## Link flags
 
     env.Append(LINKFLAGS=['-s', 'BINARYEN=1'])
-    env.Append(LINKFLAGS=['-s', 'BINARYEN_TRAP_MODE=\'clamp\''])
+    #env.Append(LINKFLAGS=['-s', 'BINARYEN_TRAP_MODE=\'clamp\''])
 
     # Allow increasing memory buffer size during runtime. This is efficient
     # when using WebAssembly (in comparison to asm.js) and works well for
```

**Comment by Houkime** [comment]
Duplicate symbol error happens on llvm 8.0.0 as well.

**Comment by Houkime** [comment]
Can one mask link flags from commandline to make it adjustable when calling scons?

**Comment by akien-mga** [comment]
You can call `scons LINKFLAGS="your custom flags"`. I'm not sure if they would end up before or after the ones added in `detect.py` though.

**Comment by Houkime** [comment]
I've managed to compile an HTML5 template using my system-wide LLVM 10 (from llvm git, LLVM_ROOT=/usr/bin) and the patch to disable BINARYEN_TRAPMODE.

However it only works if one calls scons -c before build and otherwise ends up with controllermappings issues.

**Comment by santouits** [comment]
in your command there are two time the same file, is it supposed to be like that?
``main/default_controller_mappings.gen.javascript.opt.bc``
```
shared:ERROR: '/home/akien/Projects/godot/emscripten/emsdk/upstream/bin/wasm-ld -o main/libmain.javascript.opt.bc --allow-undefined --import-memory --import-table --lto-O0 main/default_controller_mappings.gen.javascript.opt.bc main/input_default.javascript.opt.bc main/main.javascript.opt.bc main/main_timer_sync.javascript.opt.bc main/performance.javascript.opt.bc main/default_controller_mappings.gen.javascript.opt.bc -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --export __wasm_call_ctors --export __data_end --export main --export malloc --export free --export setThrew --export __errno_location --export fflush -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024 --relocatable' failed (1)
```

**Comment by akien-mga** [comment]
@santouits Good catch. Might be due to https://github.com/godotengine/godot/blob/16288cabc0fa959f4f23d7861708ad751b2c6736/main/SCsub#L17

We don't use direct `append` normally but `env.add_source_files()`, I'll see if that fixes it.

**Comment by akien-mga** [comment]
Fixed by #30751 and #30752.

**Comment by akien-mga** [comment]
> I've managed to compile an HTML5 template using my system-wide LLVM 10 (from llvm git, LLVM_ROOT=/usr/bin) and the patch to disable BINARYEN_TRAPMODE.

So you're not running into #30756? As far as I understand Emscripten's `latest-upstream` is still LLVM 9.0.0, so maybe it's a bug there which was fixed in LLVM 10~svn?
